src/core/handle.ml

```diff
@@ -258,50 +258,100 @@ let handle_cmd : sig_state -> p_command -> sig_state * proof_data option =
-      let (ss, ind_sym) = handle_symbol ss e Injec Eager id [] a in
-      (* Add the constructors in the signature.  *)
-      let add_cons :
-            sig_state * sym list -> ident * p_term -> sig_state * sym list
-        = fun (ss, cons_list) (id, a) ->
-        let (ss, cons_sym) = handle_symbol ss e Const Eager id [] a in
-        (ss, cons_sym::cons_list)
+      (* STEP 1 - Add the inductive types in the signature *)
+      let add_ind_types :
+        sig_state * sym list -> p_inductive -> sig_state * sym list
+        = fun (ss, ind_typ_list) x ->
+        let (id, a, _) = x.elt in
+        let (ss, ind_sym) = handle_symbol ss e Injec Eager id [] a in
+        (ss, ind_sym::ind_typ_list)
+      in
+      let (ss, ind_typ_list_rev) = List.fold_left add_ind_types (ss, []) il in
+      (* STEP 2 - Add the constructors in the signature. *)
+      let add_constructors :
+        sig_state * sym list list -> p_inductive -> sig_state * sym list list
+        = fun (ss, cons_list_list) x ->
+        let (_, _, l) = x.elt in
+        let add_cons :
+              sig_state * sym list -> ident * p_term -> sig_state * sym list
+          = fun (ss, cons_list) (id, a) ->
+          let (ss, cons_sym) = handle_symbol ss e Const Eager id [] a in
+          (ss, cons_sym::cons_list)
+        in
+        let (ss, cons_list_rev) = List.fold_left add_cons (ss, []) l in
+        (* Reverse the list of constructors previously computed to preserve
+           the initial order *)
+        let cons_list = List.rev cons_list_rev in
+        (ss, cons_list::cons_list_list)
+      in
+      let (ss, cons_list_list_rev) =
+        List.fold_left add_constructors (ss, []) il
       in
-      let (ss, cons_list_rev) = List.fold_left add_cons (ss, []) l in
-      (* Reverse the list of constructors previously computed to preserve the
-         initial order *)
-      let cons_list = List.rev cons_list_rev  in
-      (* Compute the induction principle *)
-      let rec_typ = Inductive.gen_rec_type ss cmd.pos ind_sym cons_list in
-      (* Check the type of the induction principle *)
-      Inductive.check_type_ind cmd.pos rec_typ;
-      (* Add the induction principle in the signature *)
-      let rec_name = Pos.make cmd.pos (Parser.add_prefix "ind_" id.elt) in
-      if Sign.mem ss.signature rec_name.elt then
-        fatal cmd.pos "Symbol [%s] already exists." rec_name.elt;
-      if StrSet.mem id.elt !(ss.signature.sign_idents) then
-        Sign.add_ident ss.signature rec_name.elt;
-      let (ss, rec_sym) =
-        Sig_state.add_symbol ss e Defin Eager rec_name rec_typ [] None
+      let ind_list = List.combine_rev ind_typ_list_rev cons_list_list_rev in
+      (* STEP 3 - Generate the induction principle. *)
+        (* A - Compute the induction principle *)
+      let rec_typ_list_rev, assoc_predicates =
+        Inductive.gen_rec_type ss cmd.pos ind_list
+      in
+      let check_and_add rec_typ id ss =
```

Comment:
check_and_add rec_typ id ss ->
`add_recursor : Sig_state.t -> sym -> term -> Sig_state.t * sym = fun ss ind_sym rec_typ ->`

Reply:
Why `recursor`? Why only `add`? I check the type too.

Reply:
So what? It is similar to the previous add functions (add_constructors, etc.).

Reply:
Ok it's better.